src/stream.ts

```diff
@@ -281,54 +46,334 @@ export enum StreamStates {
+     // through a source.
+     this._src = pipe(
+       this._innersrc,
+       lengthPrefixed.decode(),
+       (source) => (async function * () {
+         for await (const buf of source) {
+           const message = self.processIncomingProtobuf(buf.subarray())
+           if (message != null) {
+             yield new Uint8ArrayList(message)
+           }
+         }
+       })()
+     )
+   }
+
+   // If user attempts to set a new source
+   // this should be a nop
+   set source (_src: Source<Uint8ArrayList>) {
+   }
+
+   get source (): Source<Uint8ArrayList> {
+     return this._src
+   }
+
+   private async _sinkFn (src: Source<Uint8ArrayList | Uint8Array>): Promise<void> {
+     await this.opened.promise
+     if (this.streamState.state === StreamStates.CLOSED || this.streamState.state === StreamStates.WRITE_CLOSED) {
+       return
+     }
+
+     const self = this
+     const closeWriteIterable = {
+       async * [Symbol.asyncIterator] () {
+         await self.closeWritePromise.promise
+         yield new Uint8Array(0)
+       }
+     }
+
+     for await (const buf of merge(closeWriteIterable, src)) {
+       const state = self.streamState.state
+       if (state === StreamStates.CLOSED || state === StreamStates.WRITE_CLOSED) {
+         return
+       }
+       const msgbuf = pb.Message.toBinary({ message: buf.subarray() })
+       const sendbuf = lengthPrefixed.encode.single(msgbuf)
+       this.channel.send(sendbuf.subarray())
+     }
+   }
+
+   processIncomingProtobuf (buffer: Uint8Array): Uint8Array | undefined {
+     const message = pb.Message.fromBinary(buffer)
+
+     if (message.flag !== undefined) {
+       const [currentState, nextState] = this.streamState.transition({ direction: 'inbound', flag: message.flag })
+       if (currentState !== nextState) {
+         // @TODO(ddimaria): determine if we need to check for StreamStates.OPEN
+         switch (nextState) {
+           case StreamStates.READ_CLOSED:
+             this._innersrc.end()
+             break
+           case StreamStates.WRITE_CLOSED:
+             this.closeWritePromise.resolve()
+             break
+           case StreamStates.CLOSED:
+             this.close()
+             break
+
+             // no default
+         }
+       }
+     }
+     return message.message
+   }
+
+   /**
+    * Close a stream for reading and writing
+    */
+   close (): void {
+     this.stat.timeline.close = new Date().getTime()
+     this.streamState.state = StreamStates.CLOSED
+     this._innersrc.end()
+     this.closeWritePromise.resolve()
+     this.channel.close()
+
+     if (this.closeCb !== undefined) {
+       this.closeCb(this)
+     }
+   }
+
+   /**
+    * Close a stream for reading only
+    */
+   closeRead (): void {
+     const [currentState, nextState] = this.streamState.transition({ direction: 'outbound', flag: pb.Message_Flag.STOP_SENDING })
+
+     if (currentState === StreamStates.OPEN || currentState === StreamStates.WRITE_CLOSED) {
```

Comment:
lines like this definitely use a comment on why this is what it is

Reply:
I agree.  This all looks like new code but isn't.  I envision variables or other mechanisms for spelling out the intention in the next PR.  For this one, it might look like:
```
const isClosed = this.streamState.state === StreamStates.CLOSED || this.streamState.state === StreamStates.WRITE_CLOSED;

if (isClosed) {
  return;
}
```

Reply:
Again, this is all for an subsequent PR.  The intent of the PR is primarily lints, but will want a robust review on the all of the code on the next one.